### PR TITLE
Make it easy to create GANs

### LINKED ISSUE #535: Adding GAN support to TensorGraph
I think our current model with `TensorGraph` can't nicely support GANs. In order to train GANs, you need to train both a discriminator `D` and a generator `G`. The training of these two models is threaded together, with each updated for a few minibatches at a time. While updating the generator, the discriminator weights should be frozen. While updating the discriminator, the generator weights need to be frozen.

I have an idea for a new `TensorGraph` API to support this sort of training. We add a new `layers` kwarg to `fit`:

- `model.fit(feed_generator, layers=None)`

The idea is that `fit` can be instructed to train only some layers at a time. Note that there's now a semantic change:

- `Layer` is responsible for tracking the variables it introduces. 
  - `layer.variables` should be a list of all trainable variables created by that `Layer`.

@peastman and @lilleswing Does this basic design make sense? 

Hat tip to @enf for explaining GAN training to me in detail!

### PR DESCRIPTION
**Pull request by peastman**
Fixes #535.  This is my initial implementation.  I'd really appreciate feedback on the API.  I still need to create an example notebook, and I'd also like to define a subclass for WGANs.

I've tried to encode some of the best practices from https://github.com/soumith/ganhacks into this class.  As a result, it's a bit different from the CGAN notebook I wrote before.  For example, it makes sure that each batch contains either only training data, or only generated data.  That's important if you want to use batch normalization.  I also might add some sort of label noise, but I need to figure out which one is best to use.

### CODE REVIEW

**Comment by rbharath** [comment]
@peastman I like the API in general! Having to set the data/noise shape explicitly is a little awkward, but that might be unavoidable.

@joegomes, @evanfeinberg It would be great to get your feedback since you've been working with GANs. Does this API make sense to you?

**Comment by peastman** [comment]
I've been trying to create a DCGAN example, but I'm running into problems.  If I use a fully connected network for the generator and a convolutional network for the discriminator, I get reasonable results.  But if I try to use a deconvolutional network for the generator, it crashes.  I've tracked it down to a Tensorflow bug, which I've just reported: https://github.com/tensorflow/tensorflow/issues/13643.

Sigh.

**Comment by peastman** [comment]
The more I look at this, the more I think we really need to create multiple copies of the discriminator.  That's easy enough to do, except that the copies all need to share variables.  Currently the only layer that supports sharing variables is Dense.  The straightforward but labor intensive solution is to implement `shared()` for all other layers anyone might potentially want to use in a GAN.  Alternatively, it might be possible to create a generic routine that can clone any layer while preserving its variables, but I'm not sure about that.

**Comment by rbharath** [comment]
If we can go for a generic clone, that would be very nice! Someone asked me about implementing siamese networks with shared weights, but gave up because they couldn't figure out how to implement custom `shared()` ops.

On a more specific note though, why do we need multiple discriminators?

**Comment by peastman** [comment]
The reason other people don't run into the bug mentioned above is that they have two copies of the discriminator, one for training data and one for generated data.  I instead implemented it as two separate training steps, one with 0 noise inputs and `batch_size` data inputs, then a separate one with `batch_size` noise inputs and 0 data inputs.  Duplicating the discriminator would work around the bug.

But I also realized I didn't implement the gradient penalty for WGANs quite correctly.  To do that right, you actually need a third copy of the discriminator to evaluate the gradients at randomly chosen points midway between training samples and generated samples.

The easiest implementation might be a `SharedLayer` class.  When you call `shared()` on any layer, no matter what class it is, it would return a `SharedLayer`.  And that class would implement `create_tensor()` to clone the original layer's portion of the graph.

**Comment by peastman** [comment]
I think I'm going to have to give up on the `Shared` layer.  It doesn't work correctly, and I can't figure out why.  In simple tests it seems to work fine, but when I create a stack of convolutional layers, the result at the end comes out wrong.  Somehow extra information is leaking through between the original and copied layers that shouldn't.

So we're back to needing every layer to implement `shared()`.  Currently Dense is the only layer that supports it.  I'll implement it for the convolutional layers, and create a default implementation that works for any layer that doesn't create variables.  But for other layers, it will need to be implemented before they can be used in the discriminator of a GAN.

**Comment by peastman** [comment]
If people think the code looks ok, this should be ready to merge.  There's still lots to do on GANs, but that can wait for future PRs.

**Comment by rbharath** [comment]
@peastman Awesome work on this! Going to be a very useful tool.

LGTM
